### PR TITLE
Fish histfile

### PR DESCRIPTION
**Pull request by tomassedovic**
## Description

This rebases @krader1961's commit from last year onto master:

https://github.com/krader1961/fish-shell/commit/469dd55950b5d18e78ac92097eedf5bb5cd43336

and implements the `none` and `fish` special history filenames. With these changes, the user can set a custom location for the history file or disable it altogether.

Apologies to @krader1961 if this is bad manners, but I'd really like to see this feature in and I'm happy to help with the work.

While this does work for, I haven't done a thorough testing yet because I'm not sure whether people are okay with this going in.

Fixes issue #102

## TODOs:

- [x] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

**Comment by krader1961** [comment]
No need to apologize, @tomassedovic. We love to see people contribute changes to the core C++ code. I have an incomplete fix for that issue in my local git repo but keep abandoning it to work on other issues that seem to me to be more important. So it doesn't bother me in the slightest that you have decided to fix that issue. I'll review your change ASAP. Having said that, you should have added a comment to issue #102 that you were working on a fix for the issue just in case I, or someone else, was actively working on a fix for it.

File: doc_src/history.txt [hunk 1]
```diff
@@ -36,12 +37,18 @@ These flags can appear before or immediately after one of the sub-commands liste
 
 - `-e` or `--exact` searches or deletes items in the history that exactly match the specified text string. This is the default for the `--delete` flag. Note that the match is case-insensitive by default. If you really want an exact match, including letter case, you must use the `-C` or `--case-sensitive` flag.
 
+- `--file` changes the current history file. That option can be used both independent of any other option and in conjunction with the other options. In both cases the change to the current history file is persistent and affects subsequent commands. The switch to a new history file occurs before any other option is acted on and an implicit `history --save` occurs before the switch.
```
Comment (faho) [inline doc_src/history.txt]: This does not seem to work. `history --file banana` will just ignore the "--file" and search for "banana" instead.

Additionally, I'd prefer it if there was just one way to change the history file, and the variable seems to be nicer since it can also be set from outside (env FISH_HISTFILE=none fish).

File: doc_src/history.txt [hunk 2]
```diff
@@ -36,12 +37,18 @@ These flags can appear before or immediately after one of the sub-commands liste
...
+
+The "history_file_identifier" can be a simple word or a relative or absolute path name. A simple word (i.e., one without slashes) is interpreted as a file name relative to `$XDG_DATA_HOME/fish` (`$HOME/.local/share/fish` by default). If "history_file_identifier" is the special word "fish" the default history path name will be substituted. The special word "none" results in no history file being used. This is analogous to the incognito mode provided by web browsers like Chrome. A relative path (one that contains a slash but does not begin with a slash) is interpreted relative to `$PWD` at the time the command is run. The resulting path must refer to a regular file (or not exist). Special file names like /dev/null are not allowed.
```
Comment (faho) [inline doc_src/history.txt]: This wording seems overly complicated. I'd use e.g. 'If it is "fish", the default file will be used' and 'If it is "none", no history will be saved (like the incognito mode in web browsers)'.

I'd also ditch the explanation for what a relative path is.

Also, what happens if the resulting path exists but isn't a regular file? And which "Special file names" are not allowed? Just "/dev/null"?

File: doc_src/history.txt [hunk 3]
```diff
@@ -36,12 +37,18 @@ These flags can appear before or immediately after one of the sub-commands liste
...
+
 - `-p` or `--prefix` searches or deletes items in the history that begin with the specified text string. This is not currently supported by the `--delete` flag.
 
 - `-t` or `--show-time` prepends each history entry with the date and time the entry was recorded . By default it uses the strftime format `# %c%n`. You can specify another format; e.g., `--show-time='%Y-%m-%d %H:%M:%S '` or `--show-time='%a%I%p'`. The short option, `-t` doesn't accept a stftime format string; it only uses the default format. Any strftime format is allowed, including `%s` to get the raw UNIX seconds since the epoch. Note that `--with-time` is also allowed but is deprecated and will be removed at a future date.
 
 - `-z` or `--null` causes history entries written by the search operations to be terminated by a NUL character rather than a newline. This allows the output to be processed by `read -z` to correctly handle multiline history entries.
 
+If `--search` is specified without `--contains` or `--prefix`, `--contains` will be assumed.
```
Comment (faho) [inline doc_src/history.txt]: This seems unrelated to the change at hand.

File: doc_src/history.txt [hunk 4]
```diff
@@ -60,6 +67,16 @@ history delete --prefix "foo"
 # You can select more than one entry by entering their IDs seperated by a space.
 \endfish
 
+\subsection setting-the-histfile Setting the name of the history file
+
+By default interactive commands are logged to `$XDG_DATA_HOME/fish/fish_history` (typically `~/.local/share/fish/fish_history`).
+
+You can set the `FISH_HISTFILE` variable to a path name to change where commands are logged. It does not have to be an environment variable (despite the all uppercase name) and normally should not be an environment variable. You can change it at will within an interactive fish session to change where subsequent commands are logged. If an environment variable exists by that name when the shell starts then that file will be used rather than the default. Simply unset the variable to revert to the default history file.
```
Comment (faho) [inline doc_src/history.txt]: I'd use "exported variable" instead of "environment variable".

**Comment by faho** [comment]
So, I've got some comments on the design here:

- I don't really like the "magic" special values - the "fish" behavior is just the default, and the "none" behavior seems reasonable for an empty-but-set variable.

- There are two ways to do this - `history --file` and `set FISH_HISTFILE` - I'd like just one, and the variable seems more versatile (e.g. `env FISH_HISTFILE=none fish` to start an incognito session).

- The relative-path behavior seems to be thoroughly confusing and the documentation even acknowledges that - is this actually useful? I'd assume if someone sets a relative path, they don't care where the file is stored, they just want to name the session.

So, how about this instead:

- No `history --file`. One way to do it.
- If the variable is unset, use the default.
- If it is set but empty, store nothing.
- If it is set to an absolute path, use that.
- If it is set to a relative path, use it relative to the history directory (e.g. "banana" stores history in ~/.local/share/fish/banana)
- If the value is in any way invalid, e.g if the path exists but isn't a regular file or if there is more than one (`set FISH_HISTFILE foo bar`), print an error.

**Comment by zanchey** [comment]
I'm not sure about the need to support a path - I'd suggest just a session identifier.

**Comment by krader1961** [comment]
> There are two ways to do this - history --file and set FISH_HISTFILE - I'd like just one, and the variable seems more versatile (e.g. env FISH_HISTFILE=none fish to start an incognito session).

@faho, My thinking for allowing two ways was that one would be consistent with the proposal for changing the file used by the `read` command. The env var would normally only be used when launching a new fish shell. However, it may be preferable to have both history files be controlled solely by env vars.

**Comment by krader1961** [comment]
@faho, I also agree with all of your other points. I think the only point you made which might need more discussion is the use of a `history` flag to allow changing the history file on the fly for consistency with the `read` command. However, I am now leaning pretty strongly to using a env var for both cases. Since a) changing the file should be rare and b) the env var approach allows both files to be specified when a new fish shell is started as well as dynamically while running.

File: doc_src/history.txt (hunk 2 above)
Comment (krader1961) [inline doc_src/history.txt]: The phrase "special file names" in this context means any existing path that is not a regular file. For example, */dev/null*, is a character special device. The intent is that fish will `stat()` the path and if it exists and is not a regular file refuse to use it. How best to explain that is TBD.

File: doc_src/history.txt (hunk 3 above)
Comment (krader1961) [inline doc_src/history.txt]: It is probably because @tomassedovic used my very moldy work-in-progress change as their starting point and incorrectly resolved some merge conflicts. My original change does not have that text but it does appear in subsequent commits.

File: doc_src/history.txt (hunk 4 above)
Comment (krader1961) [inline doc_src/history.txt]: @faho, I agree that using "exported" is preferable to "environment" in this context. But note that the current fish docs are very inconsistent as to whether `exported` or `environment` is preferred. Do `grep 'environment var' doc_src/* | wc -l` and `grep 'exported var' doc_src/* | wc -l` and notice the former matches 48 lines while the latter matches 8 lines. 😄  The author of this PR was just copying text from my initial change to address issue #102 so he's only parroting my confusion as to the preferred nomenclature. In fact, one of our FAQ sections is titled "How do I set or clear an environment variable?".

**Comment by krader1961** [comment]
@tomassedovic, I have no objection to you basing your change on my work-in-progress change. Including copying my changes to the documentation. But in this case you should not include my original commit that was never merged upstream and did not receive any discussion. I do appreciate your attempt to retain the attribution to myself. It is just that in this instance doing so will cause confusion since my tentative change was never meant to be merged.

**Comment by faho** [comment]
Okay, so one issue with not having any special values is that it would be impossible to revert to the default just for one invocation:

```fish
env FISH_HISTFILE="" fish
# in the child
env FISH_HISTFILE=fish fish
```

This would leave the middle fish without history and the top one with the default history file. Without the "fish" special value (though I'd call it "default"),  you'd need to use "env -u" to avoid changing the middle fish's history file, which [isn't in POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/env.html). It is in the FreeBSD and GNU versions though.

I'm not sure if we can simply ignore this edge case.

File: doc_src/history.txt (hunk 2 above)
Comment (faho) [inline doc_src/history.txt]: Note that if we used "session identifiers" instead of paths (i.e. dropped support for absolute paths, only allowing filenames in the history directory), /dev/null would be impossible to use anyway.

Also I'm assuming symlinks are also invalid?

Comment (krader1961) [inline doc_src/history.txt]: Symlinks are meant to be invisible and thus valid. What matters is what the link refers to. The intent is that `stat()` is used not `lstat()`. Thus returning information about the object pointed to by the symlink.

I like the idea of using "session identifiers" rather than strings which may be relative or absolute paths. Using that idiom means `read --mode-name=$name` will continue to behave as it does today. Although possibly with some tweaks to disallow characters like slashes in the identifier. My original thinking was that interpreting the value as a path provided more flexibility. An absolute path such as */dev/null* would be exactly that. Whereas a relative path would be relative to the `$XDG_DATA_HOME` root. But that is going to cause more confusion than value. Better to interpret the value as a session ID as you propose, @faho, and allow a few special values such as "fish" (for the default file) and "none" or the empty string to mean no history should be used.

**Comment by krader1961** [comment]
> This would leave the middle fish without history and the top one with the default history file. Without the "fish" special value (though I'd call it "default"), ....

The point of using "fish" is that it is simply incorporated into the constructed path name and that is currently the default value. We might want to also special case "default" as a synonym for "fish" but that probably requires more discussion.

As for whether `env FISH_HISTFILE="" fish` means use the default history file or no history file is debatable. My preference is for it to mean no history file. That is because a value of "fish" (and possibly "default") means to use the default history file. Which leaves the question of how to specify no history should be saved or used. We can certainly special case a value like "none". But then what do you do for non-english speakers for whom "none" is meaningless? Do we need to normalize the case of the string when doing that comparison so that "NONE" also means not to write to or use a history file?

**Comment by tomassedovic** [comment]
Thanks for the reviews!

I will take a deeper look later on, but after skimming this, your suggestions make sense. 

@krader1961 I will do a rebase and thank you in the commit message instead of keeping the commit as is, then.

Regarding the potential duplication of effort: I didn't want to promise something I wasn't sure I could deliver. So I tried to just rebase your commit to see how hard would that be (turns out not very) and then added the `none` and `fish` values. All in all it took maybe 2 hours or so. If it turned out someone else was working on a better solution, I'd be fine with that. I'll comment on the issue now, though.

**Comment by tomassedovic** [comment]
To sum up, we seem to have a consensus on:

* no modifications to the `history` command interface
* if `FISH_HISTFILE` is unset, use the default history location
* if `FISH_HISTFILE` is set to an empty string, do not record history
* otherwise, the value is not a path, but a session ID that can be represented as a file (so say only alphanumeric characters plus underscore and dash allowed)

I'd like to have at most one special value and that is `default`. To fit @krader1961's design, we would treat any other non-empty value it as the prefix in `PREFIX_history`, i.e.:

`FISH_HISTFILE="fish"` -> `~/.local/share/fish/fish_history{,.*}`
`FISH_HISTFILE="banana"` -> `~/.local/share/fish/banana_history{,.*}`

(so in this case `fish` is nothing special, it just happens to expand to the current location)

Alternatively, we could say that the default value for `FISH_HISTFILE` is `fish_history` and behave like so:

`FISH_HISTFILE="fish_history"` -> `~/.local/share/fish/fish_history{,.*}`
`FISH_HISTFILE="banana"` -> `~/.local/share/fish/banana{,.*}`
`FISH_HISTFILE="fish"` -> `~/.local/share/fish/fish{,.*}`

(which is what @faho suggested)

Either approach works for me although I like the second option a little more.

Is this something everyone is happy with? And if so, which of the options for the session ID would you prefer?

**Comment by krader1961** [comment]
@tomassedovic, Your summation looks good. I feel strongly that the session ID should be a prefix rather than the entire file name. That's because we'll want to also use it in constructing the default `read` history file name. I recommend copying your summation to the associated issue. Primarily because these types of design discussions don't really belong in a code review. But also because someone searching for this discussion in the future is unlikely to include pull-requests in the set of documents they search.

**Comment by krader1961** [comment]
Also, I just checked and cannot find a single script in the fish project or oh-my-fish or fisherman which uses the `read --mode` or `read -m` flags. All the more reason to remove that flag and simply rely on the `FISH_HISTFILE` var to control which history files are used. Actually making that change is orthogonal to this PR but I mention it because it buttresses my argument for using the var as a prefix which affects both files and should be the only way to change which history files are used.

@tomassedovic, You'll also want to hook that var into the `react_to_variable_change()` function so that it can be changed on the fly.

**Comment by tomassedovic** [comment]
Thanks @krader1961!

**Comment by tomassedovic** [comment]
Alright, this implements the session ID. Would appreciate any comments (there's still a ton that I'm missing about the codebase).

Also, a question: how do we print an error message to the user? There's a TODO for that in the code (when the session id contains invalid characters).

**Comment by tomassedovic** [comment]
Hm, looks like setting the environment variable from the outside doesn't work.

**Comment by faho** [comment]
>Also, a question: how do we print an error message to the user? There's a TODO for that in the code (when the session id contains invalid characters).

Debuglevel 0 is for errors, 1 is for warnings. So this should be 0.

`debug(0, _(L"yourmessagehere"))` (where the "_" function is for translations)

File: src/history.cpp [hunk 5]
```diff
@@ -1811,7 +1811,7 @@ wcstring history_session_id() {
     } else if (valid_var_name(session_id)) {
         result = session_id;
     } else {
-        debug(0, "Wrong characters in the session ID: '%ls'. Using `fish`.",
+        debug(0, _(L"Wrong characters in the session ID: '%ls'. Using `fish`."),
```
Comment (krader1961) [inline src/history.cpp]: Something like "History session IDs must be valid fish variable names" would be more helpful. Simply telling the user they included "wrong characters" doesn't help them correct the error.

File: doc_src/index.hdr.in [hunk 6]
```diff
@@ -846,6 +846,8 @@ The user can change the settings of `fish` by changing the values of certain var
 
 - `history`, an array containing the last commands that were entered.
 
+- `FISH_HISTFILE`: Name of the history session. It defaults to `$XDG_DATA_HOME/fish/fish_history` (typically `~/.local/share/fish/fish_history`). This variable can be changed by the user. It does not have to be an environment variable. You can change it at will within an interactive fish session to change where subsequent commands are logged.
```
Comment (krader1961) [inline doc_src/index.hdr.in]: Looks like you forgot too update this message. The default should be "fish" not the entire path.

File: doc_src/index.hdr.in [hunk 7]
```diff
@@ -1129,7 +1131,7 @@ History searches can be aborted by pressing the escape key.
 
 Prefixing the commandline with a space will prevent the entire line from being stored in the history.
 
-The history is stored in the file `~/.local/share/fish/fish_history` (or `$XDG_DATA_HOME/fish/fish_history` if that variable is set).
+The command history is stored in the file `~/.local/share/fish/fish_history` (or `$XDG_DATA_HOME/fish/fish_history` if that variable is set) by default. However, you can set the `FISH_HISTFILE` environment variable to change the file name; both before starting the shell and while the shell is running.
```
Comment (krader1961) [inline doc_src/index.hdr.in]: Not sure how to rephrase this but it incorrectly implies that FISH_HISTORY controls the absolute path name.

File: src/env.cpp [hunk 8]
```diff
@@ -610,6 +610,9 @@ static void react_to_variable_change(const wcstring &key) {
         invalidate_termsize(true);  // force fish to update its idea of the terminal size plus vars
     } else if (key == L"FISH_READ_BYTE_LIMIT") {
         env_set_read_limit();
+    } else if (key == L"FISH_HISTFILE") {
+        history_destroy();
+        reader_push(history_session_id().c_str());
```
Comment (krader1961) [inline src/env.cpp]: Shouldn't there be a corresponding `reader_pop()`?

File: src/history.cpp [hunk 9]
```diff
@@ -1778,6 +1800,25 @@ void history_sanity_check() {
...
+    const env_var_t session_id = env_get_string(L"FISH_HISTFILE");
+    if (session_id.missing() || session_id.compare(L"default") == 0) {
+        result = L"fish";
+    } else if (session_id.empty()) {
+        result = L"";
+    } else if (valid_var_name(session_id)) {
+        result = session_id;
+    } else {
+        // TODO: the session id is invalid. Print an error.
+		debug(1, "TODO ERROR: Wrong characters in the session ID: '%ls'. Falling back to `fish`.", session_id.c_str());
```
Comment (krader1961) [inline src/history.cpp]: Why not just "History session ID '%ls' is not a valid var name. Falling back to `fish`."? Also, this should be localized by wrapping it in the magic `_()` function.

Comment (krader1961) [inline src/history.cpp]: Also, I would restructure this for the common case which is to use the default session ID. Start with `wcstring result = L"fish";` and only set it to something else if the var has been set and its value is valid.

**Comment by tomassedovic** [comment]
@krader1961 I've updated the error message and resolved the conflicts with master.

**Comment by tomassedovic** [comment]
Um, I somehow managed to completely miss all the other comments. Sorry! Will update them soonish.

File: src/env.cpp (hunk 8 above)
Comment (tomassedovic) [inline src/env.cpp]: I'm not sure. I mean, yeah it looks like it, but where should that pop be? To my knowledge, there's no way to "pop" the history session -- you just set it to a new value which may or not be the previous session.

We can pop the previous reader before pushing the new one here, but I don't know enough about the reader stack to know whether that's the right thing to do.

**Comment by tomassedovic** [comment]
Alright, this should address all the comments. See my question regarding the reader_push/pop suggestion.

**Comment by krader1961** [comment]
@tomassedovic: I'll try to review and test your changes this week. As for the missing `reader_pop()` it appears that should be done before the push you introduced. But hold off on doing that until I've had another chance to review and test the current state of this PR.

**Comment by tomassedovic** [comment]
Thanks, I really appreciate you taking the time with this.

**Comment by zanchey** [comment]
Having thought about this some more, I can forsee complaints that we don't allow a full pathname as the session identifier. Allowing full paths doesn't feel very fishy - all your configuration and data should be in one place, controlled by the same variable - but I'm wondering if that's a strong enough reason.

**Comment by tomassedovic** [comment]
I like the session IDs idea more -- paths can be complicated (symlinks, special files, etc). But we should be able to add support for paths later backwards compatibly: if the FISH_HISTFILE is a valid var name, we'd treat as a `<prefix>_history` and if it has slashes it would be treated as a path.

So that work could be deferred to if and when we decide to do it later.

**Comment by krader1961** [comment]
In addition to @tomassedovic's comments which I agree with I want to point out that I'm hoping to see us eliminate the `read --mode-name` flag. AFAICT no one is using it. At least there weren't any references to it in any of the OMF or Fisherman repositories. Nor can I find any fish issues that mention it outside of the two specifically dealing with making the history files customizable.

**Comment by krader1961** [comment]
I noticed that when doing `env FISH_HISTFILE=test1 fish` the resulting fish history file is populated with entries from *~/.bash_history*. This is correct but surprising given the primary purpose of this feature. I'm pretty sure we want to suppress reading, or at least discard, the bash history if upon initializing the fish session we find that `FISH_HISTFILE` is set.

Also, it's going to take a bit of work to merge this in light of the builtin refactoring and wgetopt cleanup I did in recent weeks. Since that extra work is my fault I'm willing to deal with it but that will screw up the attribution. Your choice.

**Comment by krader1961** [comment]
Squash merged as aec097319646.

**Comment by tomassedovic** [comment]
I've just got some time to look at this and saw this got merged. Thanks! And thanks for the HISTFILE and bash import fixes.
